src/Avalonia.Base/Rendering/Composition/CompositionDrawingSurface.cs

```diff
@@ -25,6 +25,19 @@ public Task UpdateWithKeyedMutexAsync(ICompositionImportedGpuImage image, uint a
         return Compositor.InvokeServerJobAsync(() => Server.UpdateWithKeyedMutex(img, acquireIndex, releaseIndex));
     }
 
+    /// <summary>
+    /// Updates the surface contents using an imported memory image using an external keyed mutex as the means of synchronization
+    /// </summary>
+    /// <param name="image">GPU image with new surface contents</param>
+    /// <param name="acquire">The acquire callback to call before accessing the image</param>
+    /// <param name="release">The release callback to call after accessing the image </param>
+    /// <returns>A task that completes when update operation is completed and user code is free to destroy or dispose the image</returns>
+    public Task UpdateWithExternalKeyedMutexAsync(ICompositionImportedGpuImage image, Action acquire, Action release)
```

Comment:
I believe that it won't work properly.

The mutex controls resources on the GPU side and should be locked/released as a part of the GPU command stream, not on the command issuer side. If you simply lock/unlock in the renderer, it won't affect the enqueued Vulkan commands, those would still be submitted to the queue without locks.

At most you'll stall the CPU.

Please, use proper Vulkan synchronization primitives that are capable of using DXGI Keyed Mutexes.